src/engine/MultiColumnJoin.cpp

```diff
@@ -0,0 +1,471 @@
+      result->_resultTypes.push_back(rightResult->_resultTypes[col]);
+    }
+  }
+
+  LOG(DEBUG) << "Computing optional join between results of size "
+             << leftResult->size() << " and " << rightResult->size() << endl;
+
+  // Calls computeMultiColumnJoin with the right values for the array sizes.
+  MultiColumnJoinCaller<1, 1, 1>::call(
+      leftResult->_nofColumns, rightResult->_nofColumns, result->_nofColumns,
+      leftResult, rightResult, _joinColumns, result, result->_nofColumns);
+  result->finish();
+  LOG(DEBUG) << "MultiColumnJoin result computation done." << endl;
+}
+
+// _____________________________________________________________________________
+ad_utility::HashMap<string, size_t> MultiColumnJoin::getVariableColumns()
+    const {
+  ad_utility::HashMap<string, size_t> retVal(_left->getVariableColumnMap());
+  size_t leftSize = _left->getResultWidth();
+  for (auto it = _right->getVariableColumnMap().begin();
+       it != _right->getVariableColumnMap().end(); ++it) {
+    size_t columnIndex = leftSize + it->second;
+    bool isJoinColumn = false;
+    // Reduce the index for every column of _right that is beeing joined on,
+    // and the index of which is smaller than the index of it.
+    for (const std::array<Id, 2>& a : _joinColumns) {
+      if (a[1] < it->second) {
+        columnIndex--;
+      } else if (a[1] == it->second) {
+        isJoinColumn = true;
+        break;
+      }
+    }
+    if (!isJoinColumn) {
+      retVal[it->first] = columnIndex;
+    }
+  }
+  return retVal;
+}
+
+// _____________________________________________________________________________
+size_t MultiColumnJoin::getResultWidth() const {
+  size_t res =
+      _left->getResultWidth() + _right->getResultWidth() - _joinColumns.size();
+  AD_CHECK(res > 0);
+  return res;
+}
+
+// _____________________________________________________________________________
+vector<size_t> MultiColumnJoin::resultSortedOn() const {
+  std::vector<size_t> sortedOn;
+  // The result is sorted on all join columns from the left subtree.
+  for (const auto& a : _joinColumns) {
+    sortedOn.push_back(a[0]);
+  }
+  return sortedOn;
+}
+
+// _____________________________________________________________________________
+float MultiColumnJoin::getMultiplicity(size_t col) {
+  if (!_multiplicitiesComputed) {
+    computeSizeEstimateAndMultiplicities();
+  }
+  return _multiplicities[col];
+}
+
+// _____________________________________________________________________________
+size_t MultiColumnJoin::getSizeEstimate() {
+  if (!_multiplicitiesComputed) {
+    computeSizeEstimateAndMultiplicities();
+  }
+  return _sizeEstimate;
+}
+
+// _____________________________________________________________________________
+size_t MultiColumnJoin::getCostEstimate() {
+  size_t costEstimate =
+      getSizeEstimate() + _left->getSizeEstimate() + _right->getSizeEstimate();
+  // The optional join is about 3-7 times slower than a normal join, due to
```

Comment:
this isn't an optional join, so is this still the right estimate?

Reply:
I reduced the factor to 2. If we want to benchmark this join and compare it to the normal `Join` operation anyways we can then use that data to get a better factor.